File: python/eventgraph/bus.py
"""Event bus — pub/sub fan-out with non-blocking delivery."""

from __future__ import annotations

import threading
from typing import Callable

from .event import Event
from .store import InMemoryStore
from .types import EventType, SubscriptionPattern


class EventBus:
    """Thread-safe event bus with per-subscriber buffering.

    Slow subscribers get dropped events, not blocked publishers.
    """

    def __init__(self, store: InMemoryStore, buffer_size: int = 256) -> None:
        self._store = store
        self._buffer_size = max(buffer_size, 1)
        self._lock = threading.Lock()
        self._subs: dict[int, _Subscription] = {}
        self._next_id = 0
        self._closed = False

    @property
    def store(self) -> InMemoryStore:
        return self._store

    def subscribe(
        self, pattern: SubscriptionPattern, handler: Callable[[Event], None]
    ) -> int:
        """Register a handler for events matching the pattern.

        Returns subscription ID, or -1 if the bus is closed.
        """
        with self._lock:
            if self._closed:
                return -1
            self._next_id += 1
            sub_id = self._next_id
            sub = _Subscription(sub_id, pattern, handler, self._buffer_size)
            self._subs[sub_id] = sub
            sub.start()
            return sub_id

    def unsubscribe(self, sub_id: int) -> None:
        with self._lock:
            sub = self._subs.pop(sub_id, None)
        if sub is not None:
            sub.close()

    def publish(self, event: Event) -> None:
        """Deliver event to all matching subscribers. Non-blocking."""
        with self._lock:
            if self._closed:
                return
            subs = list(self._subs.values())

        for sub in subs:
            sub.deliver(event)

    def close(self, timeout: float = 30.0) -> None:
        """Stop all subscriber threads."""
        with self._lock:
            if self._closed:
                return
            self._closed = True
            subs = list(self._subs.values())
            self._subs.clear()

        for sub in subs:
            sub.close()
        for sub in subs:
            sub.join(timeout=timeout)


class _Subscription:
    """Per-subscriber delivery thread with bounded buffer."""

    def __init__(
        self,
        sub_id: int,
        pattern: SubscriptionPattern,
        handler: Callable[[Event], None],
        buffer_size: int,
    ) -> None:
        self.id = sub_id
        self.pattern = pattern
        self.handler = handler
        self._buffer: list[Event] = []
        self._buffer_size = buffer_size
        self._cond = threading.Condition()
        self._closed = False
        self._thread: threading.Thread | None = None
        self.last_error: Exception | None = None

    def start(self) -> None:
        self._thread = threading.Thread(target=self._run, daemon=True)
        self._thread.start()

    def deliver(self, event: Event) -> None:
        """Non-blocking delivery. Drops if buffer full or pattern doesn't match."""
        if not self.pattern.matches(event.type):
            return
        with self._cond:
            if self._closed:
                return
            if len(self._buffer) < self._buffer_size:
                self._buffer.append(event)
                self._cond.notify()

    def close(self) -> None:
        with self._cond:
            self._closed = True
            self._cond.notify()

    def join(self, timeout: float = 30.0) -> None:
        if self._thread is not None:
            self._thread.join(timeout=timeout)

    def _run(self) -> None:
        while True:
            with self._cond:
                while not self._buffer and not self._closed:
                    self._cond.wait()
                if self._closed and not self._buffer:
                    return
                event = self._buffer.pop(0)

            try:
                self.handler(event)
            except Exception as e:
                self.last_error = e

File: dotnet/src/EventGraph/Bus.cs
using System.Collections.Concurrent;

namespace EventGraph;

/// <summary>Thread-safe event bus with per-subscriber buffering and non-blocking delivery.</summary>
public sealed class EventBus : IDisposable
{
    private readonly InMemoryStore _store;
    private readonly int _bufferSize;
    private readonly Lock _lock = new();
    private readonly Dictionary<int, Subscription> _subs = new();
    private int _nextId;
    private bool _closed;

    public EventBus(InMemoryStore store, int bufferSize = 256)
    {
        _store = store;
        _bufferSize = Math.Max(bufferSize, 1);
    }

    public InMemoryStore Store => _store;

    public int Subscribe(SubscriptionPattern pattern, Action<Event> handler)
    {
        lock (_lock)
        {
            if (_closed) return -1;
            var id = ++_nextId;
            var sub = new Subscription(id, pattern, handler, _bufferSize);
            _subs[id] = sub;
            sub.Start();
            return id;
        }
    }

    public void Unsubscribe(int subId)
    {
        Subscription? sub;
        lock (_lock) { _subs.Remove(subId, out sub); }
        sub?.Stop();
    }

    public void Publish(Event ev)
    {
        List<Subscription> snapshot;
        lock (_lock)
        {
            if (_closed) return;
            snapshot = new(_subs.Values);
        }
        foreach (var sub in snapshot)
            sub.Deliver(ev);
    }

    public void Dispose()
    {
        List<Subscription> subs;
        lock (_lock)
        {
            if (_closed) return;
            _closed = true;
            subs = new(_subs.Values);
            _subs.Clear();
        }
        foreach (var sub in subs)
            sub.Stop();
        foreach (var sub in subs)
            sub.Join(TimeSpan.FromSeconds(30));
    }

    private sealed class Subscription
    {
        private readonly int _id;
        private readonly SubscriptionPattern _pattern;
        private readonly Action<Event> _handler;
        private readonly BlockingCollection<Event> _buffer;
        private Thread? _thread;
        public Exception? LastError;

        public Subscription(int id, SubscriptionPattern pattern, Action<Event> handler, int bufferSize)
        {
            _id = id;
            _pattern = pattern;
            _handler = handler;
            _buffer = new BlockingCollection<Event>(bufferSize);
        }

        public void Start()
        {
            _thread = new Thread(Run) { IsBackground = true };
            _thread.Start();
        }

        public void Deliver(Event ev)
        {
            if (!_pattern.Matches(ev.Type)) return;
            _buffer.TryAdd(ev); // drops if full
        }

        public void Stop() => _buffer.CompleteAdding();

        public void Join(TimeSpan timeout) => _thread?.Join(timeout);

        private void Run()
        {
            try
            {
                foreach (var ev in _buffer.GetConsumingEnumerable())
                {
                    try { _handler(ev); }
                    catch (Exception ex) { LastError = ex; }
                }
            }
            catch (OperationCanceledException) { }
        }
    }
}
